test/AftermarketDeviceId.test.ts

```diff
@@ -0,0 +1,403 @@
+      }
+    });
+    const claimOwnerSig2 = await signMessage({
+      _signer: user1,
+      _primaryType: 'ClaimAftermarketDeviceSign',
+      _verifyingContract: aftermarketDeviceInstance.address,
+      message: {
+        aftermarketDeviceNode: '2',
+        owner: user1.address
+      }
+    });
+    const claimAdSig2 = await signMessage({
+      _signer: adAddress2,
+      _primaryType: 'ClaimAftermarketDeviceSign',
+      _verifyingContract: aftermarketDeviceInstance.address,
+      message: {
+        aftermarketDeviceNode: '2',
+        owner: user1.address
+      }
+    });
+    const pairSignature = await signMessage({
+      _signer: user1,
+      _primaryType: 'PairAftermarketDeviceSign',
+      _verifyingContract: aftermarketDeviceInstance.address,
+      message: {
+        aftermarketDeviceNode: '1',
+        vehicleNode: '1'
+      }
+    });
+
+    await vehicleInstance
+      .connect(admin)
+      .mintVehicle(1, user1.address, C.mockVehicleAttributeInfoPairs);
+    await aftermarketDeviceInstance
+      .connect(admin)
+      .claimAftermarketDeviceSign(
+        1,
+        user1.address,
+        claimOwnerSig1,
+        claimAdSig1
+      );
+    await aftermarketDeviceInstance
+      .connect(admin)
+      .claimAftermarketDeviceSign(
+        2,
+        user1.address,
+        claimOwnerSig2,
+        claimAdSig2
+      );
+    await aftermarketDeviceInstance
+      .connect(admin)
+      .pairAftermarketDeviceSign(1, 1, pairSignature);
+  });
+
+  beforeEach(async () => {
+    snapshot = await createSnapshot();
+  });
+
+  afterEach(async () => {
+    await revertToSnapshot(snapshot);
+  });
+
+  describe('setDimoRegistryAddress', () => {
+    it('Should revert if addr is zero address', async () => {
+      await expect(
+        adIdInstance
+          .connect(admin)
+          .setDimoRegistryAddress(C.ZERO_ADDRESS)
+      ).to.be.revertedWith('Non zero address');
+    });
+  })
+
+  context('On transfer', async () => {
+    context('Error handling', () => {
+      it('Should revert if caller is not the token owner or DimoRegistry', async () => {
+        await expect(
+          adIdInstance
+            .connect(user2)['safeTransferFrom(address,address,uint256)'](user1.address, user2.address, 1)
+        ).to.be.revertedWith('ERC721: caller is not token owner nor approved');
+      });
+      it('Should revert if aftermarket device is already paired', async () => {
+        await expect(
+          adIdInstance
+            .connect(user1)['safeTransferFrom(address,address,uint256)'](user1.address, user2.address, 1)
+        ).to.be.revertedWith('AD must not be paired');
+      });
+    });
+
+    context('State', () => {
+      it('Should keep the same parent node', async () => {
+        const parentNode = await nodesInstance.getParentNode(adIdInstance.address, 2);
+
+        await adIdInstance.connect(user1)['safeTransferFrom(address,address,uint256)'](user1.address, user2.address, 2);
```

Comment:
There's some reason why you have to do this and for the life of me I can't remember what it is.

To be honest I wouldn't have thought to test these registry properties. This is good.

Reply:
Bc we have two `safeTransferFrom(address,address,uint256)` and `safeTransferFrom(address,address,uint256, bytes)`. So we cannot do simply `contract.safeTransferFrom`, we have to specify which one we want.

Reply:
Oh yeah overloading ahahaha